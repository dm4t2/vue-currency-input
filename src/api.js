import createCurrencyFormat from './utils/createCurrencyFormat'
import parse from './utils/parse'

/**
 * Parses a number from a currency formatted string.
 * @param {String} formattedValue The currency formatted string to be parsed, for example `$1,234.50`.
 * @param {String} locale A [BCP 47](https://tools.ietf.org/html/bcp47) language tag (for example `en` or `de-DE`).
 * @param {String | Object} currency A [ISO 4217](https://en.wikipedia.org/wiki/ISO_4217) currency code (for example `USD` or `EUR`), an object `{prefix, suffix}` or `null`.
 * @param {Boolean} valueAsInteger Use `true` to restore the decimal digits of a formatted value, when using the `valueAsInteger` option of the component/directive.
 * @returns {number | null} The parsed number or `null` if the formatted string does not match.
 */
export const parseCurrency = (formattedValue, locale, currency, valueAsInteger) => parse(formattedValue, createCurrencyFormat({ locale, currency }), valueAsInteger)
