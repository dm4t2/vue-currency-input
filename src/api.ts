import { Ref } from 'vue'

/**
 * @internal
 */
export interface CurrencyInputValue {
  number: number | null
  formatted: string | null
}

export interface NumberRange {
  min?: number
  max?: number
}

export enum CurrencyDisplay {
  symbol = 'symbol',
  narrowSymbol = 'narrowSymbol',
  code = 'code',
  name = 'name',
  hidden = 'hidden'
}

export enum ValueScaling {
  precision = 'precision',
  thousands = 'thousands',
  tenThousands = 'tenThousands',
  millions = 'millions',
  billions = 'billions'
}

export interface CurrencyInputOptions {
  accountingSign?: boolean
  autoDecimalDigits?: boolean
  currency: string
  currencyDisplay?: CurrencyDisplay
  hideCurrencySymbolOnFocus?: boolean
  hideGroupingSeparatorOnFocus?: boolean
  hideNegligibleDecimalDigitsOnFocus?: boolean
  locale?: string
  precision?: NumberRange | number
  useGrouping?: boolean
  valueRange?: NumberRange
  valueScaling?: ValueScaling
}

export interface UseCurrencyInput {
  formattedValue: Ref<string | null>
  inputRef: Ref
  numberValue: Ref<number | null>
  setOptions: (options: CurrencyInputOptions) => void
  setValue: (number: number | null) => void
}
