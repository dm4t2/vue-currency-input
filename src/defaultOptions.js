export default {
  locale: undefined,
  currency: 'EUR',
  valueAsInteger: false,
  distractionFree: true,
  decimalLength: undefined,
  autoDecimalMode: false,
  min: null,
  max: null
}
