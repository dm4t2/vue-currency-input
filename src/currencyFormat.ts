import { escapeRegExp, substringBefore } from './utils'
import { CurrencyDisplay, CurrencyInputOptions } from './api'
import NumberFormatOptions = Intl.NumberFormatOptions

export const DECIMAL_SEPARATORS = [',', '.', 'Ù«']
export const INTEGER_PATTERN = '(0|[1-9]\\d*)'

export default class CurrencyFormat {
  locale?: string
  currency: string
  currencyDisplay: CurrencyDisplay | undefined
  digits: string[]
  decimalSymbol: string | undefined
  groupingSymbol: string | undefined
  minusSymbol: string
  minimumFractionDigits: number
  maximumFractionDigits: number
  prefix: string
  negativePrefix: string
  suffix: string

  constructor(options: CurrencyInputOptions) {
    const { currency, currencyDisplay, locale, precision } = options
    this.currencyDisplay = currencyDisplay !== CurrencyDisplay.hidden ? currencyDisplay : undefined
    const numberFormat = new Intl.NumberFormat(locale, { currency, currencyDisplay: this.currencyDisplay, style: 'currency' })
    const formatSample = numberFormat.format(1)
    const formatParts = numberFormat.formatToParts(123456)

    this.locale = locale
    this.currency = currency
    this.digits = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9].map((i) => i.toLocaleString(locale))
    this.decimalSymbol = formatParts.find(({ type }) => type === 'decimal')?.value
    this.groupingSymbol = formatParts.find(({ type }) => type === 'group')?.value
    this.minusSymbol = substringBefore(Number(-1).toLocaleString(locale), this.digits[1])

    if (this.decimalSymbol === undefined) {
      this.minimumFractionDigits = this.maximumFractionDigits = 0
    } else if (typeof precision === 'number') {
      this.minimumFractionDigits = this.maximumFractionDigits = precision
    } else {
      this.minimumFractionDigits = precision?.min ?? numberFormat.resolvedOptions().minimumFractionDigits
      this.maximumFractionDigits = precision?.max ?? numberFormat.resolvedOptions().maximumFractionDigits
    }

    this.prefix = substringBefore(formatSample, this.digits[1])
    this.negativePrefix = substringBefore(numberFormat.format(-1), this.digits[1])
    this.suffix = formatSample.substring(formatSample.lastIndexOf(this.decimalSymbol ? this.digits[0] : this.digits[1]) + 1)
  }

  parse(str: string | null): number | null {
    if (str) {
      const negative = this.isNegative(str)
      str = this.normalizeDigits(str)
      str = this.stripCurrencySymbol(str, negative)
      str = this.stripMinusSymbol(str)
      const fraction = this.decimalSymbol ? `(?:${escapeRegExp(this.decimalSymbol)}(\\d*))?` : ''
      const match = this.stripGroupingSeparator(str).match(new RegExp(`^${INTEGER_PATTERN}${fraction}$`))
      if (match && this.isValidIntegerFormat(this.decimalSymbol ? str.split(this.decimalSymbol)[0] : str, Number(match[1]))) {
        return Number(`${negative ? '-' : ''}${this.onlyDigits(match[1])}.${this.onlyDigits(match[2] || '')}`)
      }
    }
    return null
  }

  isValidIntegerFormat(formattedNumber: string, integerNumber: number): boolean {
    const options = { style: 'currency', currency: this.currency, currencyDisplay: this.currencyDisplay, minimumFractionDigits: 0 }
    return [
      this.stripCurrencySymbol(
        this.normalizeDigits(
          integerNumber.toLocaleString(this.locale, {
            ...options,
            useGrouping: true
          })
        ),
        false
      ),
      this.stripCurrencySymbol(
        this.normalizeDigits(
          integerNumber.toLocaleString(this.locale, {
            ...options,
            useGrouping: false
          })
        ),
        false
      )
    ].includes(formattedNumber)
  }

  format(
    value: number | null,
    options: NumberFormatOptions = {
      minimumFractionDigits: this.minimumFractionDigits,
      maximumFractionDigits: this.maximumFractionDigits
    }
  ): string {
    return value != null
      ? value.toLocaleString(this.locale, {
          style: 'currency',
          currency: this.currency,
          currencyDisplay: this.currencyDisplay,
          ...options
        })
      : ''
  }

  toFraction(str: string): string {
    return `${this.digits[0]}${this.decimalSymbol}${this.onlyLocaleDigits(str.substr(1)).substr(0, this.maximumFractionDigits)}`
  }

  isFractionIncomplete(str: string): boolean {
    return !!this.normalizeDigits(this.stripGroupingSeparator(str)).match(new RegExp(`^${INTEGER_PATTERN}${escapeRegExp(this.decimalSymbol as string)}$`))
  }

  isNegative(str: string): boolean {
    return str.startsWith(this.negativePrefix) || str.replace('-', this.minusSymbol).startsWith(this.minusSymbol)
  }

  insertCurrencySymbol(str: string, negative: boolean): string {
    return `${negative ? this.negativePrefix : this.prefix}${str}${this.suffix}`
  }

  stripGroupingSeparator(str: string): string {
    return this.groupingSymbol !== undefined ? str.replace(new RegExp(escapeRegExp(this.groupingSymbol), 'g'), '') : str
  }

  stripMinusSymbol(str: string): string {
    return str.replace('-', this.minusSymbol).replace(this.minusSymbol, '')
  }

  stripCurrencySymbol(str: string, negative: boolean): string {
    return str.replace(negative ? this.negativePrefix : this.prefix, '').replace(this.suffix, '')
  }

  normalizeDecimalSeparator(str: string, from: number): string {
    DECIMAL_SEPARATORS.forEach((s) => {
      str = str.substr(0, from) + str.substr(from).replace(s, this.decimalSymbol as string)
    })
    return str
  }

  normalizeDigits(str: string): string {
    if (this.digits[0] !== '0') {
      this.digits.forEach((digit, index) => {
        str = str.replace(new RegExp(digit, 'g'), String(index))
      })
    }
    return str
  }

  onlyDigits(str: string): string {
    return this.normalizeDigits(str).replace(/\D+/g, '')
  }

  onlyLocaleDigits(str: string): string {
    return str.replace(new RegExp(`[^${this.digits.join('')}]*`, 'g'), '')
  }
}
