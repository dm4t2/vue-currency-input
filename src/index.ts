import useCurrencyInput from './useCurrencyInput'
import CurrencyFormat from './currencyFormat'
import { abs } from './utils'
import { CurrencyFormatOptions } from './api'

export const parseBigInt = (formattedValue: string, options: CurrencyFormatOptions) => new CurrencyFormat(options).parse(formattedValue)

export const parseNumber = (formattedValue: string, options: CurrencyFormatOptions): number | null => {
  const currencyFormat = new CurrencyFormat(options)
  const value = currencyFormat.parse(formattedValue)
  if (value != null) {
    if (abs(value) <= Number.MAX_SAFE_INTEGER) {
      let number = Number(value)
      if (currencyFormat.maximumFractionDigits) {
        number /= 10 ** currencyFormat.maximumFractionDigits
      }
      return number
    } else {
      throw new Error('Value exceeds Number.MAX_SAFE_INTEGER')
    }
  }
  return null
}

export * from './api'
export { useCurrencyInput }
