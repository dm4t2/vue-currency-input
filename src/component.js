import defaultOptions from './defaultOptions'
import currencyDirective from './directive'
import dispatchEvent from './utils/dispatchEvent'

export default {
  render (h) {
    return h('input', {
      domProps: {
        value: this.formattedValue
      },
      directives: [{
        name: 'currency',
        value: this.options
      }],
      on: this.listeners()
    })
  },
  directives: {
    currency: currencyDirective
  },
  name: 'CurrencyInput',
  props: {
    value: {
      type: Number,
      default: null
    },
    locale: {
      type: String,
      default: undefined
    },
    currency: {
      type: [String, Object],
      default: undefined
    },
    distractionFree: {
      type: [Boolean, Object],
      default: undefined
    },
    decimalLength: {
      type: Number,
      default: undefined
    },
    autoDecimalMode: {
      type: Boolean,
      default: undefined
    },
    valueAsInteger: {
      type: Boolean,
      default: undefined
    },
    min: {
      type: Number,
      default: undefined
    },
    max: {
      type: Number,
      default: undefined
    }
  },
  data () {
    return {
      formattedValue: this.value
    }
  },
  computed: {
    options () {
      const options = { ...this.$CI_DEFAULT_OPTIONS || defaultOptions }
      Object.keys(defaultOptions).forEach((key) => {
        if (this[key] !== undefined) {
          options[key] = this[key]
        }
      })
      return options
    }
  },
  watch: {
    value (value) {
      dispatchEvent(this.$el, 'format', { value })
    }
  },
  methods: {
    listeners () {
      const { input, ...listeners } = this.$listeners // all but input event
      return {
        ...listeners,
        'format-complete': ({ detail }) => {
          this.$emit('input', detail.numberValue)
          this.formattedValue = this.$el.value
        }
      }
    }
  }
}
