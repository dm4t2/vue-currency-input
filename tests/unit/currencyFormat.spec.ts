import CurrencyFormat from '../../src/currencyFormat'
import { CurrencyDisplay } from '../../src'
import { describe, expect, it } from 'vitest'

describe('CurrencyFormat', () => {
  describe('constructing number formats', () => {
    it('should work for respective locale and currency', () => {
      expect(new CurrencyFormat({ locale: 'de-DE', currency: 'EUR' })).toMatchSnapshot('de-DE_EUR')
      expect(new CurrencyFormat({ locale: 'de-CH', currency: 'EUR' })).toMatchSnapshot('de-CH_EUR')
      expect(new CurrencyFormat({ locale: 'es-ES', currency: 'EUR' })).toMatchSnapshot('es-ES_EUR')
      expect(new CurrencyFormat({ locale: 'nl-NL', currency: 'EUR' })).toMatchSnapshot('nl-NL_EUR')
      expect(new CurrencyFormat({ locale: 'en-US', currency: 'USD' })).toMatchSnapshot('en-US_USD')
      expect(new CurrencyFormat({ locale: 'fr-CH', currency: 'CHF' })).toMatchSnapshot('fr-CH_CHF')
      expect(new CurrencyFormat({ locale: 'zh', currency: 'CNY' })).toMatchSnapshot('zh_CNY')
      expect(new CurrencyFormat({ locale: 'en-GB', currency: 'GBP' })).toMatchSnapshot('en-GB_GBP')
      expect(new CurrencyFormat({ locale: 'en-GB', currency: 'INR' })).toMatchSnapshot('en-IN_INR')
      expect(new CurrencyFormat({ locale: 'pt', currency: 'BRL' })).toMatchSnapshot('pt_BRL')
      expect(new CurrencyFormat({ locale: 'ja', currency: 'JPY' })).toMatchSnapshot('ja_JPY')
      expect(new CurrencyFormat({ locale: 'ar-SA', currency: 'SAR' })).toMatchSnapshot('ar-SA_SAR')
      expect(new CurrencyFormat({ locale: 'fa-IR', currency: 'IRR' })).toMatchSnapshot('fa-IR_IRR')
      expect(new CurrencyFormat({ locale: 'bg-BG', currency: 'BGN' })).toMatchSnapshot('bg-BG_BGN')
    })

    describe('custom precision', () => {
      it('should work with a custom precision', () => {
        expect(new CurrencyFormat({ currency: 'EUR', precision: 0 })).toEqual(expect.objectContaining({ minimumFractionDigits: 0, maximumFractionDigits: 0 }))
      })

      it('should work with a custom precision range', () => {
        expect(new CurrencyFormat({ currency: 'EUR', precision: { min: 0, max: 4 } })).toEqual(
          expect.objectContaining({ minimumFractionDigits: 0, maximumFractionDigits: 4 })
        )
      })

      it('should ignore the custom precision if the locale does not support decimal digits', () => {
        expect(new CurrencyFormat({ locale: 'ja', currency: 'JPY', precision: 2 })).toEqual(
          expect.objectContaining({ minimumFractionDigits: 0, maximumFractionDigits: 0 })
        )
      })
    })
  })

  describe('parse', () => {
    it('should return null if the value is empty or invalid', () => {
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR' }).parse(null)).toBeNull()
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR' }).parse('')).toBeNull()
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR' }).parse(' ')).toBeNull()
      expect(new CurrencyFormat({ currency: 'EUR', locale: 'en' }).parse('-')).toBeNull()
    })

    it('should return the parsed number if the value conforms to the currency format', () => {
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR' }).parse('1234')).toBe(123400n)
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR' }).parse('1,234,567')).toBe(123456700n)
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR' }).parse('-1,234,567')).toBe(-123456700n)
      expect(new CurrencyFormat({ locale: 'de', currency: 'EUR' }).parse('-1234567,89')).toBe(-123456789n)
      expect(new CurrencyFormat({ locale: 'en', currency: 'USD' }).parse('$1,234,567')).toBe(123456700n)
      expect(new CurrencyFormat({ locale: 'de', currency: 'EUR' }).parse('1234 €')).toBe(123400n)
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR' }).parse('-1234')).toBe(-123400n)
      expect(new CurrencyFormat({ locale: 'en', currency: 'USD' }).parse('-$1234')).toBe(-123400n)
      expect(new CurrencyFormat({ locale: 'de', currency: 'EUR' }).parse('-1234 €')).toBe(-123400n)
      expect(new CurrencyFormat({ locale: 'ja', currency: 'JPY' }).parse('￥123,456')).toBe(123456n)
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR' }).parse('0.5')).toBe(50n)
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR' }).parse('1234.50')).toBe(123450n)
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR' }).parse('1234.00')).toBe(123400n)
      expect(new CurrencyFormat({ locale: 'en', currency: 'USD' }).parse('$1,234.50')).toBe(123450n)
      expect(new CurrencyFormat({ locale: 'de', currency: 'EUR' }).parse('1.234,50 €')).toBe(123450n)
      expect(new CurrencyFormat({ locale: 'ar', currency: 'SAR' }).parse('١٢٣٤')).toBe(123400n)
      expect(new CurrencyFormat({ locale: 'ar', currency: 'SAR' }).parse('١٬٢٣٤')).toBe(123400n)
      expect(new CurrencyFormat({ locale: 'ar', currency: 'SAR' }).parse('٠٫٩')).toBe(90n)
      expect(new CurrencyFormat({ locale: 'ar', currency: 'SAR' }).parse('-‏٠٫٥٠ ر.س.')).toBe(-50n)
      expect(new CurrencyFormat({ locale: 'en-IN', currency: 'INR' }).parse('₹1,23,334.00')).toBe(12333400n)
      expect(new CurrencyFormat({ locale: 'en-IN', currency: 'INR' }).parse('₹123334.00')).toBe(12333400n)
      expect(new CurrencyFormat({ locale: 'de-AT', currency: 'EUR' }).parse('€ 66.668')).toBe(6666800n)
      expect(new CurrencyFormat({ locale: 'de-DE', currency: 'USD', currencyDisplay: CurrencyDisplay.name }).parse('1.234,50 US-Dollar')).toBe(123450n)
      expect(new CurrencyFormat({ locale: 'en', currency: 'USD', accountingSign: true }).parse('(1,234.50)')).toBe(-123450n)
    })
  })

  describe('isFractionIncomplete', () => {
    it('should return true if the fraction is incomplete', () => {
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR' }).isFractionIncomplete('1234.')).toBe(true)
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR' }).isFractionIncomplete('1234')).toBe(false)
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR' }).isFractionIncomplete('1234.3')).toBe(false)
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR' }).isFractionIncomplete('1234.3.')).toBe(false)
      expect(new CurrencyFormat({ locale: 'de', currency: 'EUR' }).isFractionIncomplete('1,3,')).toBe(false)
    })
  })

  describe('normalizeDecimalSymbol', () => {
    it('should replace the first decimal symbol with the one of the current locale', () => {
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR' }).normalizeDecimalSeparator('1,23,4,567', 2)).toBe('1,23.4,567')
    })
  })

  describe('format', () => {
    it('should return the formatted value for the respective options', () => {
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR' }).format(null)).toBe('')
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR' }).format(1n)).toBe('€0.01')
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR' }).format(100n, { minimumFractionDigits: 0, maximumFractionDigits: 2 })).toBe('€1')
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR' }).format(0n, { minimumFractionDigits: 1, maximumFractionDigits: 2 })).toBe('€0.0')
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR' }).format(120n, { minimumFractionDigits: 1, maximumFractionDigits: 2 })).toBe('€1.2')
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR' }).format(100n, { minimumFractionDigits: 0, maximumFractionDigits: 0 })).toBe('€1')
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR' }).format(12345789n)).toBe('€123,457.89')
      expect(new CurrencyFormat({ locale: 'en', currency: 'JPY' }).format(12345789n)).toBe('¥12,345,789')
    })

    it('should apply the custom precision range', () => {
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR', precision: { min: 0, max: 4 } }).format(1234n)).toBe('€0.1234')
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR', precision: { min: 0, max: 4 } }).format(12345n)).toBe('€1.2345')
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR', precision: { min: 0, max: 4 } }).format(123457n)).toBe('€12.3457')
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR', precision: { min: 0, max: 4 } }).format(1234578n)).toBe('€123.4578')
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR', precision: { min: 0, max: 4 } }).format(12345789n)).toBe('€1,234.5789')
      expect(new CurrencyFormat({ locale: 'en', currency: 'EUR', precision: { min: 0, max: 4 } }).format(123457891n)).toBe('€12,345.7891')
    })
  })

  describe('isNegative', () => {
    it('should check if a formatted value is negative or not', () => {
      expect(new CurrencyFormat({ locale: 'en', currency: 'USD', accountingSign: true }).isNegative('($1)')).toBe(true)
      expect(new CurrencyFormat({ locale: 'en', currency: 'USD', accountingSign: true }).isNegative('(1)')).toBe(true)
      expect(new CurrencyFormat({ locale: 'en', currency: 'USD', accountingSign: true }).isNegative('-1')).toBe(true)
      expect(new CurrencyFormat({ locale: 'sv', currency: 'EUR' }).isNegative('-1.')).toBe(true)
    })
  })
})
