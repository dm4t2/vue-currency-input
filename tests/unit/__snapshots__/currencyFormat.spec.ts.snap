// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CurrencyFormat > constructing number formats > should work for respective locale and currency > ar-SA_SAR 1`] = `
CurrencyFormat {
  "currency": "ر.س.",
  "decimalSymbol": "٫",
  "digits": [
    "٠",
    "١",
    "٢",
    "٣",
    "٤",
    "٥",
    "٦",
    "٧",
    "٨",
    "٩",
  ],
  "groupingSymbol": "٬",
  "locale": "ar-SA",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSign": "-",
  "negativePrefix": "؜-‏",
  "negativeSuffix": " ر.س.‏",
  "options": {
    "currency": "SAR",
    "currencyDisplay": undefined,
    "currencySign": undefined,
    "style": "currency",
    "useGrouping": undefined,
  },
  "prefix": "‏",
  "suffix": " ر.س.‏",
}
`;

exports[`CurrencyFormat > constructing number formats > should work for respective locale and currency > bg-BG_BGN 1`] = `
CurrencyFormat {
  "currency": "лв.",
  "decimalSymbol": ",",
  "digits": [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": " ",
  "locale": "bg-BG",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSign": "-",
  "negativePrefix": "-",
  "negativeSuffix": " лв.",
  "options": {
    "currency": "BGN",
    "currencyDisplay": undefined,
    "currencySign": undefined,
    "style": "currency",
    "useGrouping": undefined,
  },
  "prefix": "",
  "suffix": " лв.",
}
`;

exports[`CurrencyFormat > constructing number formats > should work for respective locale and currency > de-CH_EUR 1`] = `
CurrencyFormat {
  "currency": "EUR",
  "decimalSymbol": ".",
  "digits": [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": "’",
  "locale": "de-CH",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSign": "-",
  "negativePrefix": "EUR-",
  "negativeSuffix": "",
  "options": {
    "currency": "EUR",
    "currencyDisplay": undefined,
    "currencySign": undefined,
    "style": "currency",
    "useGrouping": undefined,
  },
  "prefix": "EUR ",
  "suffix": "",
}
`;

exports[`CurrencyFormat > constructing number formats > should work for respective locale and currency > de-DE_EUR 1`] = `
CurrencyFormat {
  "currency": "€",
  "decimalSymbol": ",",
  "digits": [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": ".",
  "locale": "de-DE",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSign": "-",
  "negativePrefix": "-",
  "negativeSuffix": " €",
  "options": {
    "currency": "EUR",
    "currencyDisplay": undefined,
    "currencySign": undefined,
    "style": "currency",
    "useGrouping": undefined,
  },
  "prefix": "",
  "suffix": " €",
}
`;

exports[`CurrencyFormat > constructing number formats > should work for respective locale and currency > en-GB_GBP 1`] = `
CurrencyFormat {
  "currency": "£",
  "decimalSymbol": ".",
  "digits": [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": ",",
  "locale": "en-GB",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSign": "-",
  "negativePrefix": "-£",
  "negativeSuffix": "",
  "options": {
    "currency": "GBP",
    "currencyDisplay": undefined,
    "currencySign": undefined,
    "style": "currency",
    "useGrouping": undefined,
  },
  "prefix": "£",
  "suffix": "",
}
`;

exports[`CurrencyFormat > constructing number formats > should work for respective locale and currency > en-IN_INR 1`] = `
CurrencyFormat {
  "currency": "₹",
  "decimalSymbol": ".",
  "digits": [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": ",",
  "locale": "en-GB",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSign": "-",
  "negativePrefix": "-₹",
  "negativeSuffix": "",
  "options": {
    "currency": "INR",
    "currencyDisplay": undefined,
    "currencySign": undefined,
    "style": "currency",
    "useGrouping": undefined,
  },
  "prefix": "₹",
  "suffix": "",
}
`;

exports[`CurrencyFormat > constructing number formats > should work for respective locale and currency > en-US_USD 1`] = `
CurrencyFormat {
  "currency": "$",
  "decimalSymbol": ".",
  "digits": [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": ",",
  "locale": "en-US",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSign": "-",
  "negativePrefix": "-$",
  "negativeSuffix": "",
  "options": {
    "currency": "USD",
    "currencyDisplay": undefined,
    "currencySign": undefined,
    "style": "currency",
    "useGrouping": undefined,
  },
  "prefix": "$",
  "suffix": "",
}
`;

exports[`CurrencyFormat > constructing number formats > should work for respective locale and currency > es-ES_EUR 1`] = `
CurrencyFormat {
  "currency": "€",
  "decimalSymbol": ",",
  "digits": [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": ".",
  "locale": "es-ES",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSign": "-",
  "negativePrefix": "-",
  "negativeSuffix": " €",
  "options": {
    "currency": "EUR",
    "currencyDisplay": undefined,
    "currencySign": undefined,
    "style": "currency",
    "useGrouping": undefined,
  },
  "prefix": "",
  "suffix": " €",
}
`;

exports[`CurrencyFormat > constructing number formats > should work for respective locale and currency > fa-IR_IRR 1`] = `
CurrencyFormat {
  "currency": "ریال",
  "decimalSymbol": undefined,
  "digits": [
    "۰",
    "۱",
    "۲",
    "۳",
    "۴",
    "۵",
    "۶",
    "۷",
    "۸",
    "۹",
  ],
  "groupingSymbol": "٬",
  "locale": "fa-IR",
  "maximumFractionDigits": 0,
  "minimumFractionDigits": 0,
  "minusSign": "−",
  "negativePrefix": "‎−‎ریال ",
  "negativeSuffix": "",
  "options": {
    "currency": "IRR",
    "currencyDisplay": undefined,
    "currencySign": undefined,
    "style": "currency",
    "useGrouping": undefined,
  },
  "prefix": "‎ریال ",
  "suffix": "",
}
`;

exports[`CurrencyFormat > constructing number formats > should work for respective locale and currency > fr-CH_CHF 1`] = `
CurrencyFormat {
  "currency": "CHF",
  "decimalSymbol": ".",
  "digits": [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": " ",
  "locale": "fr-CH",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSign": "-",
  "negativePrefix": "-",
  "negativeSuffix": " CHF",
  "options": {
    "currency": "CHF",
    "currencyDisplay": undefined,
    "currencySign": undefined,
    "style": "currency",
    "useGrouping": undefined,
  },
  "prefix": "",
  "suffix": " CHF",
}
`;

exports[`CurrencyFormat > constructing number formats > should work for respective locale and currency > ja_JPY 1`] = `
CurrencyFormat {
  "currency": "￥",
  "decimalSymbol": undefined,
  "digits": [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": ",",
  "locale": "ja",
  "maximumFractionDigits": 0,
  "minimumFractionDigits": 0,
  "minusSign": "-",
  "negativePrefix": "-￥",
  "negativeSuffix": "",
  "options": {
    "currency": "JPY",
    "currencyDisplay": undefined,
    "currencySign": undefined,
    "style": "currency",
    "useGrouping": undefined,
  },
  "prefix": "￥",
  "suffix": "",
}
`;

exports[`CurrencyFormat > constructing number formats > should work for respective locale and currency > nl-NL_EUR 1`] = `
CurrencyFormat {
  "currency": "€",
  "decimalSymbol": ",",
  "digits": [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": ".",
  "locale": "nl-NL",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSign": "-",
  "negativePrefix": "€ -",
  "negativeSuffix": "",
  "options": {
    "currency": "EUR",
    "currencyDisplay": undefined,
    "currencySign": undefined,
    "style": "currency",
    "useGrouping": undefined,
  },
  "prefix": "€ ",
  "suffix": "",
}
`;

exports[`CurrencyFormat > constructing number formats > should work for respective locale and currency > pt_BRL 1`] = `
CurrencyFormat {
  "currency": "R$",
  "decimalSymbol": ",",
  "digits": [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": ".",
  "locale": "pt",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSign": "-",
  "negativePrefix": "-R$ ",
  "negativeSuffix": "",
  "options": {
    "currency": "BRL",
    "currencyDisplay": undefined,
    "currencySign": undefined,
    "style": "currency",
    "useGrouping": undefined,
  },
  "prefix": "R$ ",
  "suffix": "",
}
`;

exports[`CurrencyFormat > constructing number formats > should work for respective locale and currency > zh_CNY 1`] = `
CurrencyFormat {
  "currency": "¥",
  "decimalSymbol": ".",
  "digits": [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": ",",
  "locale": "zh",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSign": "-",
  "negativePrefix": "-¥",
  "negativeSuffix": "",
  "options": {
    "currency": "CNY",
    "currencyDisplay": undefined,
    "currencySign": undefined,
    "style": "currency",
    "useGrouping": undefined,
  },
  "prefix": "¥",
  "suffix": "",
}
`;
